### DISCUSSION TITLE
Continuation prompt (PS2)

### DISCUSSION THREAD

**Discussion by JanDeDobbeleer** [chrisant996/clink#254] [Q&A] [answered]
I was going through the docs but couldn't immediately find if there's a prompt filter available for a continuation prompt. We recently added that functionality to [oh-my-posh](https://ohmyposh.dev/docs/config-secondary-prompt) so I was looking at porting this to clink as well.

**Comment by chrisant996** [comment, marked as answer]
That's correct, Clink currently doesn't support customizing the continuation prompt, and doesn't enable any of its input line editing support in the continuation prompt.

I haven't taken any steps down that path because:
1. There isn't a reliable enough way to detect when CMD shows the continuation prompt.  Currently Clink uses the combination of seeing a particular string, plus having additional lines already queued up, and then it auto-responds with pretty high confidence to a continuation prompt.  But when there aren't already lines queued up, there isn't a reliable enough way detect whether it's a really continuation prompt.
2. Clink also doesn't understand how to lex and parse CMD syntax for the `(` and `)` grouping symbols, and doesn't understand any CMD control statements (`if`, `for`, and so on).  It gets very complicated very quickly (and expensive, and likely unreliable).  I don't currently plan to teach Clink about any of those.  So, adding more support for fancy continuation prompt behaviors would likely set a misleading precedent.

**Reply by JanDeDobbeleer** [reply]
Thanks, just wanted to make sure I didn't miss anything obvious 🙏🏻

**Comment by chrisant996** [comment]
The question got my subconscious thinking about it:

1. I realized there's already a problem with `( echo foo )`:  Clink thinks `(` is the command and doesn't handle argmatchers correctly.
    - Item 1 is worth addressing no matter what.
2. I realized parsing the `(` and `)` grouping symbols can take some shortcuts.
    - Item 2 is worth doing to solve item 1, and also to support coloring the `(` and `)` appropriately (instead of as commands or arguments).
3. I realized a way to trick CMD and reliably detect the continuation prompt.
    - I will likely hook up Clink's input line to work in the continuation prompt.
    - I may add a way to filter the continuation prompt, but I haven't decided yet.

**Reply by JanDeDobbeleer** [reply]
This response is so relatable 😄 I can't even count the times where I made an elaborate statement on something not being feasible only to come to a solution a day or so later. Let me know how it evolves and if I can somehow hook into it 🙏🏻

**Reply by chrisant996** [reply]
The shortcuts I hoped to take for parsing the `(` and `)` grouping symbols don't really work. Proper parsing looks very complicated and requires special handling (and quirks) for `rem`, `if`, `for`, etc.

Items 1 and 3 still look feasible.

Item 2 is more involved:
- Handling leading `(` is feasible.
- Handling parsing open/close levels for parentheses is feasible.
  - But `rem` messes that up because it eats the whole rest of the line.
  - But `rem` already messes up `&` and `|` parsing.
  - A special exception can be made for `rem`.
- Truly accurate handling of `(` and `)` syntax -- and syntax errors -- is very complicated.
- I'll look for a way to apply the "80 / 20 rule". I don't think it's worth the cost of being entirely/pedantically accurate.

**Comment by chrisant996** [comment]
The following makes the continuation prompt super complicated to handle in any kind of reasonable way:

```
C:\somewhere> echo hello wor^
More? _
```

That's why I had decided not to even try to support the continuation prompt.

So, all that I'm going to do is:
- Handle the `rem` issue.
- Skip past leading `(` when looking up an argmatcher for the command.
  - Except when the `(` is part of a doskey macro name.
- Resolve doskey macros in `( macro )` when `doskey.enhanced` is enabled.
- ~~_Maybe:_ I might add coloring for `(` and `)`, and possibly even add coloring for some syntax errors related to `(` and `)` usage.~~

No continuation prompt support.
